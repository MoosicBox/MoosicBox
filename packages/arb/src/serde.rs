use quickcheck::{Arbitrary, Gen};
use serde_json::Value;

use crate::xml::XmlString;

#[derive(Clone, Debug)]
pub struct JsonValue(pub Value);

impl Arbitrary for JsonValue {
    fn arbitrary(g: &mut Gen) -> Self {
        Self(Value::String(XmlString::arbitrary(g).0))
    }
}

#[derive(Clone, Debug)]
pub struct JsonF64(pub f64);

impl Arbitrary for JsonF64 {
    fn arbitrary(g: &mut Gen) -> Self {
        loop {
            let num = f64::arbitrary(g);

            if !num.is_finite() {
                continue;
            }

            return Self(num);
        }
    }
}

#[derive(Clone, Debug)]
pub struct JsonF32(pub f32);

impl Arbitrary for JsonF32 {
    fn arbitrary(g: &mut Gen) -> Self {
        loop {
            let num = f32::arbitrary(g);

            if !num.is_finite() {
                continue;
            }

            return Self(num);
        }
    }
}
