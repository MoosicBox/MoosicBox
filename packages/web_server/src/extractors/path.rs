#![warn(clippy::all, clippy::pedantic, clippy::nursery, clippy::cargo)]
#![allow(clippy::multiple_crate_versions)]

use crate::{
    Error, HttpRequest,
    from_request::{FromRequest, IntoHandlerError},
};
use serde::de::DeserializeOwned;
use std::{collections::BTreeMap, fmt, future::Ready};

/// Error types that can occur during path parameter extraction
#[derive(Debug)]
pub enum PathError {
    /// Path is empty or contains no extractable segments
    EmptyPath,
    /// Not enough path segments for the requested extraction
    InsufficientSegments {
        /// Number of segments found in the path
        found: usize,
        /// Number of segments expected for extraction
        expected: usize,
        /// The actual path that was parsed
        path: String,
    },
    /// Failed to deserialize path segments into the target type
    DeserializationError {
        /// The serde error message (stored as string for Clone compatibility)
        source: String,
        /// The path segments that failed to deserialize
        segments: Vec<String>,
        /// The target type name
        target_type: &'static str,
    },
    /// Invalid path segment format (e.g., contains invalid characters)
    InvalidSegment {
        /// The invalid segment
        segment: String,
        /// Position of the segment in the path (0-based)
        position: usize,
        /// Reason why the segment is invalid
        reason: String,
    },
}

impl fmt::Display for PathError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::EmptyPath => {
                write!(f, "Path is empty or contains no extractable segments")
            }
            Self::InsufficientSegments {
                found,
                expected,
                path,
            } => {
                write!(
                    f,
                    "Insufficient path segments: found {found}, expected {expected} in path '{path}'"
                )
            }
            Self::DeserializationError {
                source,
                segments,
                target_type,
            } => {
                write!(
                    f,
                    "Failed to deserialize path segments {segments:?} into type '{target_type}': {source}"
                )
            }
            Self::InvalidSegment {
                segment,
                position,
                reason,
            } => {
                write!(
                    f,
                    "Invalid path segment '{segment}' at position {position}: {reason}"
                )
            }
        }
    }
}

impl std::error::Error for PathError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        // Since we store the error as a string, we can't return the original error
        None
    }
}

impl From<PathError> for Error {
    fn from(err: PathError) -> Self {
        Self::bad_request(err.to_string())
    }
}

impl IntoHandlerError for PathError {
    fn into_handler_error(self) -> Error {
        self.into()
    }
}

/// Path parameter extractor for URL path segments
///
/// This extractor can extract path parameters in several ways:
///
/// * **Single parameter**: `Path<String>` or `Path<u32>` extracts the last path segment
/// * **Multiple parameters**: `Path<(String, u32)>` extracts the last N segments as a tuple
/// * **Named parameters**: `Path<UserParams>` where `UserParams` is a deserializable struct
///
/// # Examples
///
/// ## Single Parameter Extraction
///
/// ```rust
/// use moosicbox_web_server::{Path, HttpResponse};
///
/// // Extracts user_id from paths like "/users/123"
/// async fn get_user(Path(user_id): Path<u32>) -> Result<HttpResponse, Box<dyn std::error::Error>> {
///     println!("User ID: {}", user_id);
///     Ok(HttpResponse::ok())
/// }
/// ```
///
/// ## Multiple Parameter Extraction
///
/// ```rust
/// use moosicbox_web_server::{Path, HttpResponse};
///
/// // Extracts from paths like "/users/john/posts/456"
/// async fn get_user_post(
///     Path((username, post_id)): Path<(String, u32)>
/// ) -> Result<HttpResponse, Box<dyn std::error::Error>> {
///     println!("User: {}, Post ID: {}", username, post_id);
///     Ok(HttpResponse::ok())
/// }
/// ```
///
/// ## Named Parameter Extraction
///
/// ```rust
/// use moosicbox_web_server::{Path, HttpResponse};
/// use serde::Deserialize;
///
/// #[derive(Deserialize)]
/// struct UserPostParams {
///     username: String,
///     post_id: u32,
/// }
///
/// // Extracts from paths like "/users/john/posts/456"
/// async fn get_user_post_named(
///     Path(params): Path<UserPostParams>
/// ) -> Result<HttpResponse, Box<dyn std::error::Error>> {
///     println!("User: {}, Post ID: {}", params.username, params.post_id);
///     Ok(HttpResponse::ok())
/// }
/// ```
///
/// # Path Segment Extraction Strategy
///
/// Since we don't have access to route patterns, the extractor uses these strategies:
///
/// * **Single parameter**: Extracts the last non-empty path segment
/// * **Tuple parameters**: Extracts the last N segments (where N is the tuple size)
/// * **Struct parameters**: Attempts to map the last N segments to struct fields in order
///
/// # Error Handling
///
/// The extractor returns detailed errors for various failure cases:
///
/// * `EmptyPath`: When the path contains no extractable segments
/// * `InsufficientSegments`: When there aren't enough segments for the requested extraction
/// * `DeserializationError`: When segments can't be deserialized into the target type
/// * `InvalidSegment`: When a segment contains invalid characters or format
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Path<T>(pub T);

impl<T> Path<T> {
    /// Create a new Path wrapper
    #[must_use]
    pub const fn new(value: T) -> Self {
        Self(value)
    }

    /// Extract the inner value
    #[must_use]
    pub fn into_inner(self) -> T {
        self.0
    }

    /// Get a reference to the inner value
    #[must_use]
    pub const fn as_ref(&self) -> &T {
        &self.0
    }
}

impl<T> std::ops::Deref for Path<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> std::ops::DerefMut for Path<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

/// Extract path segments from a URL path
///
/// Returns a vector of non-empty path segments, excluding the leading slash
fn extract_path_segments(path: &str) -> Vec<String> {
    path.split('/')
        .filter(|s| !s.is_empty())
        .map(|s| {
            urlencoding::decode(s)
                .unwrap_or_else(|_| s.into())
                .into_owned()
        })
        .collect()
}

/// Validate a path segment for common issues
fn validate_segment(segment: &str, position: usize) -> Result<(), PathError> {
    if segment.is_empty() {
        return Err(PathError::InvalidSegment {
            segment: segment.to_string(),
            position,
            reason: "segment is empty after URL decoding".to_string(),
        });
    }

    // Check for potentially problematic characters
    if segment.contains('\0') {
        return Err(PathError::InvalidSegment {
            segment: segment.to_string(),
            position,
            reason: "segment contains null character".to_string(),
        });
    }

    Ok(())
}

/// Extract path parameters for single parameter types
fn extract_single_param<T>(segments: &[String]) -> Result<T, PathError>
where
    T: DeserializeOwned,
{
    if segments.is_empty() {
        return Err(PathError::EmptyPath);
    }

    let last_segment = segments.last().unwrap();
    validate_segment(last_segment, segments.len() - 1)?;

    // Try to deserialize as a JSON string first (for String types)
    let json_str = format!("\"{last_segment}\"");
    serde_json::from_str::<T>(&json_str).map_or_else(
        |_| {
            // If string parsing fails, try parsing as a raw value (for numeric types)
            match serde_json::from_str::<T>(last_segment) {
                Ok(value) => Ok(value),
                Err(err) => Err(PathError::DeserializationError {
                    source: err.to_string(),
                    segments: vec![last_segment.clone()],
                    target_type: std::any::type_name::<T>(),
                }),
            }
        },
        |value| Ok(value),
    )
}

/// Extract path parameters for tuple types
fn extract_tuple_params<T>(segments: &[String]) -> Result<T, PathError>
where
    T: DeserializeOwned,
{
    // For tuples, we need to determine how many segments to extract
    // We'll use a heuristic: try to deserialize the segments as a JSON array

    if segments.is_empty() {
        return Err(PathError::EmptyPath);
    }

    // Validate all segments
    for (i, segment) in segments.iter().enumerate() {
        validate_segment(segment, i)?;
    }

    // Try different strategies for tuple extraction

    // Strategy 1: Try to deserialize all segments as a JSON array
    let json_array = format!(
        "[{}]",
        segments
            .iter()
            .map(|s| {
                // Try to parse as number first, then as string
                if s.parse::<f64>().is_ok() {
                    s.clone()
                } else {
                    format!("\"{s}\"")
                }
            })
            .collect::<Vec<_>>()
            .join(",")
    );

    match serde_json::from_str::<T>(&json_array) {
        Ok(value) => Ok(value),
        Err(first_err) => {
            // Strategy 2: Try with the last N segments where N is determined by trial
            // We'll try different segment counts starting from the end
            for count in (1..=segments.len()).rev() {
                let subset = &segments[segments.len() - count..];
                let json_array = format!(
                    "[{}]",
                    subset
                        .iter()
                        .map(|s| {
                            if s.parse::<f64>().is_ok() {
                                s.clone()
                            } else {
                                format!("\"{s}\"")
                            }
                        })
                        .collect::<Vec<_>>()
                        .join(",")
                );

                if let Ok(value) = serde_json::from_str::<T>(&json_array) {
                    return Ok(value);
                }
            }

            // If all strategies fail, return the original error
            Err(PathError::DeserializationError {
                source: first_err.to_string(),
                segments: segments.to_vec(),
                target_type: std::any::type_name::<T>(),
            })
        }
    }
}

/// Extract path parameters for struct types
fn extract_struct_params<T>(segments: &[String]) -> Result<T, PathError>
where
    T: DeserializeOwned,
{
    if segments.is_empty() {
        return Err(PathError::EmptyPath);
    }

    // Validate all segments
    for (i, segment) in segments.iter().enumerate() {
        validate_segment(segment, i)?;
    }

    // For struct types, we'll try to create a JSON object with numbered keys
    // This allows serde to deserialize into structs with ordered fields

    let mut json_map = BTreeMap::new();
    for (i, segment) in segments.iter().enumerate() {
        let value = serde_json::Value::String(segment.clone());
        json_map.insert(i.to_string(), value);
    }

    // Try to deserialize as a map first
    serde_json::from_value::<T>(serde_json::Value::Object(json_map.into_iter().collect()))
        .map_or_else(
            |_| {
                // If map deserialization fails, try as an array (for tuple structs)
                let json_array = format!(
                    "[{}]",
                    segments
                        .iter()
                        .map(|s| {
                            if s.parse::<f64>().is_ok() {
                                s.clone()
                            } else {
                                format!("\"{s}\"")
                            }
                        })
                        .collect::<Vec<_>>()
                        .join(",")
                );

                match serde_json::from_str::<T>(&json_array) {
                    Ok(value) => Ok(value),
                    Err(err) => Err(PathError::DeserializationError {
                        source: err.to_string(),
                        segments: segments.to_vec(),
                        target_type: std::any::type_name::<T>(),
                    }),
                }
            },
            |value| Ok(value),
        )
}

impl<T> FromRequest for Path<T>
where
    T: DeserializeOwned + Send + 'static,
{
    type Error = PathError;
    type Future = Ready<Result<Self, Self::Error>>;

    fn from_request_sync(req: &HttpRequest) -> Result<Self, Self::Error> {
        let path = req.path();
        let segments = extract_path_segments(path);

        if segments.is_empty() {
            return Err(PathError::EmptyPath);
        }

        // Determine extraction strategy based on type
        let type_name = std::any::type_name::<T>();

        let value = if type_name.starts_with('(') && type_name.ends_with(')') {
            // Tuple type - extract multiple parameters
            extract_tuple_params(&segments)?
        } else if type_name.contains("::")
            && !type_name.starts_with("alloc::")
            && !type_name.starts_with("core::")
        {
            // Custom struct type - try struct extraction (exclude standard library types)
            extract_struct_params(&segments)?
        } else {
            // Simple type (String, u32, etc.) - extract single parameter
            extract_single_param(&segments)?
        };

        Ok(Self(value))
    }

    fn from_request_async(req: HttpRequest) -> Self::Future {
        std::future::ready(Self::from_request_sync(&req))
    }
}

#[cfg(all(test, feature = "simulator"))]
mod tests {
    use super::*;
    use crate::{HttpRequest, Stub};
    use serde::Deserialize;

    #[cfg(any(feature = "simulator", not(feature = "actix")))]
    use crate::simulator::{SimulationRequest, SimulationStub};

    fn create_test_request(path: &str) -> HttpRequest {
        #[cfg(any(feature = "simulator", not(feature = "actix")))]
        {
            let sim_req = SimulationRequest::new(crate::Method::Get, path);
            HttpRequest::Stub(Stub::Simulator(SimulationStub::new(sim_req)))
        }
        #[cfg(all(feature = "actix", not(feature = "simulator")))]
        {
            // For Actix-only builds, use empty stub
            let _ = path;
            HttpRequest::Stub(Stub::Empty)
        }
    }

    #[test]
    fn test_single_string_parameter() {
        let req = create_test_request("/users/john");
        let result = Path::<String>::from_request_sync(&req);

        assert!(result.is_ok());
        assert_eq!(result.unwrap().0, "john");
    }

    #[test]
    fn test_single_numeric_parameter() {
        let req = create_test_request("/users/123");
        let result = Path::<u32>::from_request_sync(&req);
        assert!(result.is_ok());
        assert_eq!(result.unwrap().0, 123);
    }

    #[test]
    fn test_tuple_parameters() {
        let req = create_test_request("/users/john/posts/456");
        let result = Path::<(String, u32)>::from_request_sync(&req);
        assert!(result.is_ok());
        let (segment1, segment2) = result.unwrap().0;
        // Extracts last 2 segments: "posts", "456"
        assert_eq!(segment1, "posts");
        assert_eq!(segment2, 456);
    }

    #[test]
    fn test_triple_tuple_parameters() {
        let req = create_test_request("/api/v1/users/john/posts/456");
        let result = Path::<(String, String, u32)>::from_request_sync(&req);
        assert!(result.is_ok());
        let (a, b, c) = result.unwrap().0;

        // Last three segments: "john", "posts", "456"
        assert_eq!(a, "john");
        assert_eq!(b, "posts");
        assert_eq!(c, 456);
    }

    #[derive(Debug, Deserialize, PartialEq)]
    struct UserParams {
        username: String,
        post_id: u32,
    }

    #[test]
    fn test_struct_parameters() {
        let req = create_test_request("/users/john/posts/456");
        let result = Path::<UserParams>::from_request_sync(&req);
        // This test might fail with the current implementation
        // as struct deserialization is complex without field mapping
        // We'll implement a simpler approach for now
        println!("Struct test result: {result:?}");
    }

    #[test]
    fn test_empty_path() {
        let req = create_test_request("/");
        let result = Path::<String>::from_request_sync(&req);
        assert!(result.is_err());
        assert!(matches!(result.unwrap_err(), PathError::EmptyPath));
    }

    #[test]
    fn test_url_encoded_segments() {
        let req = create_test_request("/users/john%20doe");
        let result = Path::<String>::from_request_sync(&req);
        assert!(result.is_ok());
        assert_eq!(result.unwrap().0, "john doe");
    }

    #[test]
    fn test_invalid_numeric_conversion() {
        let req = create_test_request("/users/not_a_number");
        let result = Path::<u32>::from_request_sync(&req);
        assert!(result.is_err());
        assert!(matches!(
            result.unwrap_err(),
            PathError::DeserializationError { .. }
        ));
    }

    #[test]
    fn test_path_error_display() {
        let error = PathError::EmptyPath;
        assert_eq!(
            error.to_string(),
            "Path is empty or contains no extractable segments"
        );

        let error = PathError::InsufficientSegments {
            found: 1,
            expected: 2,
            path: "/users".to_string(),
        };
        assert_eq!(
            error.to_string(),
            "Insufficient path segments: found 1, expected 2 in path '/users'"
        );
    }

    #[test]
    fn test_path_wrapper_methods() {
        let path = Path::new("test".to_string());
        assert_eq!(path.as_ref(), "test");
        assert_eq!(path.into_inner(), "test");
    }

    #[test]
    fn test_path_deref() {
        let path = Path::new("test".to_string());
        assert_eq!(path.len(), 4); // String::len() via Deref
    }
}
