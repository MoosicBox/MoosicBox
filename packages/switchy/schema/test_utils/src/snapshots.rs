//! Snapshot testing utilities for migration verification using JSON format
//!
//! This module provides utilities for capturing and comparing database schemas
//! and migration results using insta's snapshot testing with JSON serialization.
//! JSON is used for its wide compatibility, active maintenance, and human readability
//! when pretty-printed.

use crate::TestError;
use std::path::PathBuf;
use switchy_database::{Database, DatabaseError};
use switchy_schema::MigrationError;

#[cfg(feature = "snapshots")]
use serde::{Deserialize, Serialize};

/// Error type for snapshot testing operations
#[derive(Debug, thiserror::Error)]
pub enum SnapshotError {
    /// Database operation failed
    #[error("Database error: {0}")]
    Database(#[from] DatabaseError),

    /// Migration operation failed
    #[error("Migration error: {0}")]
    Migration(#[from] MigrationError),

    /// IO operation failed
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),

    /// Snapshot validation failed
    #[error("Snapshot validation failed: {0}")]
    Validation(String),

    /// Test utilities error
    #[error("Test error: {0}")]
    Test(#[from] TestError),

    /// JSON serialization/deserialization error
    #[error("JSON error: {0}")]
    Json(#[from] serde_json::Error),
}

/// Result type for snapshot operations
pub type Result<T> = std::result::Result<T, SnapshotError>;

/// Snapshot structure for migration state verification
/// Note: This structure will grow in later phases.
/// Breaking changes to snapshot structure are acceptable during development.
/// Regenerate snapshots with `cargo insta review` when structure changes.
#[cfg(feature = "snapshots")]
#[derive(Debug, Serialize, Deserialize)]
struct MigrationSnapshot {
    test_name: String,
    migration_sequence: Vec<String>,
}

/// Placeholder for snapshot testing functionality
/// Full implementation will come in Phase 11.4.2+
pub struct SnapshotTester {
    // Implementation to follow in subsequent phases
}

/// Migration snapshot test struct for verifying database schema changes
pub struct MigrationSnapshotTest {
    test_name: String,
    migrations_dir: PathBuf,
    assert_schema: bool,
    assert_sequence: bool,
}

impl MigrationSnapshotTest {
    /// Create a new migration snapshot test
    #[must_use]
    pub fn new(test_name: &str) -> Self {
        Self {
            test_name: test_name.to_string(),
            // Points to dedicated snapshot test migrations
            migrations_dir: PathBuf::from("./test-resources/snapshot-migrations/minimal"),
            assert_schema: true,
            assert_sequence: true,
        }
    }

    #[must_use]
    pub fn migrations_dir(mut self, path: impl Into<PathBuf>) -> Self {
        self.migrations_dir = path.into();
        self
    }

    #[must_use]
    pub const fn assert_schema(mut self, enabled: bool) -> Self {
        self.assert_schema = enabled;
        self
    }

    #[must_use]
    pub const fn assert_sequence(mut self, enabled: bool) -> Self {
        self.assert_sequence = enabled;
        self
    }

    /// Optionally integrate with existing test builder for complex scenarios
    #[must_use]
    pub fn with_test_builder(self, _builder: crate::MigrationTestBuilder<'_>) -> Self {
        // Will be implemented in later phases
        self
    }

    /// Create a test database using existing utilities
    ///
    /// # Errors
    ///
    /// * Returns `SnapshotError` if database creation fails
    #[cfg(feature = "snapshots")]
    async fn create_test_database(&self) -> Result<Box<dyn Database>> {
        // Use existing test_utils helper (SQLite in-memory)
        // This database persists for the entire test lifecycle
        let db = crate::create_empty_in_memory()
            .await
            .map_err(TestError::from)?;
        Ok(db)
    }

    /// Run the snapshot test
    ///
    /// # Errors
    ///
    /// * Returns `SnapshotError` if test execution fails
    #[cfg(feature = "snapshots")]
    pub async fn run(self) -> Result<()> {
        // Create SQLite database - persists for entire test
        let db = self.create_test_database().await?;

        // Verify database works
        db.exec_raw("SELECT 1").await?;

        // Create snapshot with database info
        let snapshot = MigrationSnapshot {
            test_name: self.test_name.clone(),
            migration_sequence: vec![], // No migrations yet
        };

        insta::assert_json_snapshot!(self.test_name, snapshot);
        Ok(())
    }

    /// Run the snapshot test (non-snapshots version)
    ///
    /// # Errors
    ///
    /// * Returns `SnapshotError` if test execution fails
    #[cfg(not(feature = "snapshots"))]
    pub fn run(self) -> Result<()> {
        // Still minimal but uses configuration
        println!("Test: {}", self.test_name);
        println!("Migrations: {}", self.migrations_dir.display());
        println!(
            "Schema: {}, Sequence: {}",
            self.assert_schema, self.assert_sequence
        );

        Ok(())
    }
}
