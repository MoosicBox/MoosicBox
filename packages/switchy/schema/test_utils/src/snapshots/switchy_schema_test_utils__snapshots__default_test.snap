---
source: packages/switchy/schema/test_utils/src/snapshots.rs
expression: snapshot
---
{
  "test_name": "default_test",
  "migration_sequence": []
}
